[RegAllocFast] Ensure live-in vregs get reloaded after INLINEASM_BR spills

antoniofrighetto opened the pull request:
We have already ensured in 9cec2b246e719533723562950e56c292fe5dd5ad that `INLINEASM_BR` output operands get spilled onto the stack, both in the fallthrough path and in the indirect targets. Since reloads of live-ins values into physical registers contextually happens after all MIR instructions (and ops) have been visited, make sure such loads are placed at the start of the block, but after prologues or `INLINEASM_BR` spills, as otherwise this may cause stale values to be read from the stack.

Fixes: #74483, #110251.

qcolombet commented on llvm/lib/CodeGen/RegAllocFast.cpp:
```diff
@@ -491,6 +493,20 @@ static bool dominates(InstrPosIndexes &PosIndexes, const MachineInstr &A,
+      MFI->isSpillSlotObjectIndex(FI)) {
+    for (const auto &Op : MI.operands())
+      if (Op.isReg() && any_of(MBB->liveins(), [&](const auto &RegP) {
+            return Op.getReg() == RegP.PhysReg;
```
Can't we use `MBB->isLiveIn(Op.getReg())` here instead of `any_of`?

qcolombet commented on llvm/lib/CodeGen/RegAllocFast.cpp:
```diff
@@ -391,6 +391,8 @@ class RegAllocFastImpl {
   bool mayLiveOut(Register VirtReg);
   bool mayLiveIn(Register VirtReg);
 
+  bool isInlineAsmBrSpill(const MachineInstr &MI) const;
```
Two things:
1. It needs a comment because at first "I was what is a inlineasm branch spill"
2. This is a misnomer because unless I'm mistaken the instruction may not be spilling a value coming from an inlineasm branch. I would go with `mayBeSpillFromInlineAsmBr`

antoniofrighetto replied:
I had been pondering a few ones, and eventually I shortened the name too much. Updated, thanks!

antoniofrighetto committed:
[RegAllocFast] Introduce test for PR131350 (NFC)

antoniofrighetto committed:
[RegAllocFast] Ensure live-in vregs get reloaded after INLINEASM_BR spills

We have already ensured in 9cec2b246e719533723562950e56c292fe5dd5ad
that `INLINEASM_BR` output operands get spilled onto the stack, both
in the fallthrough path and in the indirect targets. Since reloads of
live-ins values into physical registers contextually happen after all
MIR instructions (and ops) have been visited, make sure such loads are
placed at the start of the block, but after prologues or `INLINEASM_BR`
spills, as otherwise this may cause stale values to be read from the
stack.

Fixes: #74483, #110251.

nickdesaulniers commented:
BIG BIG thanks for fixing this! Sorry I wasn't around for reviews...life has been hectic for me lately with the job change (and kid)!

antoniofrighetto commented:
> BIG BIG thanks for fixing this! Sorry I wasn't around for reviews...life has been hectic for me lately with the job change (and kid)!

Thanks, and glad to hear that! :)
